ocrd/ocrd/resolver.py

```diff
@@ -30,62 +34,77 @@ def download_to_directory(self, directory, url, basename=None, overwrite=False,
             directory (string): Directory to download files to
             basename (string, None): basename part of the filename on disk.
             url (string): URL to download from
-            overwrite (boolean): Whether to overwrite existing files with that name
+            if_exists (string, "skip"): Waht to do if target file already exists. One of ``skip`` (default), ``overwrite`` or ``raise``
             subdir (string, None): Subdirectory to create within the directory. Think fileGrp.
 
         Returns:
-            Local filename
+            Local filename, __relative__ to directory
         """
         log = getLogger('ocrd.resolver.download_to_directory') # pylint: disable=redefined-outer-name
-        log.debug("directory=|%s| url=|%s| basename=|%s| overwrite=|%s| subdir=|%s|", directory, url, basename, overwrite, subdir)
+        log.info("directory=|%s| url=|%s| basename=|%s| if_exists=|%s| subdir=|%s|", directory, url, basename, if_exists, subdir)
 
-        if url is None:
+        if not url:
             raise Exception("'url' must be a string")
-        if directory is None:
-            raise Exception("'directory' must be a string")
-
-        if basename is None:
-            if (subdir is not None) or \
-                (directory and url.startswith('file://%s' % directory)): # in case downloading a url 'file:///tmp/foo/bar' to directory '/tmp/foo'
-                basename = url.rsplit('/', 1)[-1]
-            else:
-                basename = safe_filename(url)
-
-        if subdir is not None:
-            basename = join(subdir, basename)
-
-        outfilename = join(directory, basename)
-
-        if exists(outfilename) and not overwrite:
-            log.debug("File already exists and overwrite=False: %s", outfilename)
-            return outfilename
-
-        outfiledir = outfilename.rsplit('/', 1)[0]
-        #  print(outfiledir)
-        if not isdir(outfiledir):
-            makedirs(outfiledir)
-
-        log.debug("Downloading <%s> to '%s'", url, outfilename)
-
-        # de-scheme file:// URL
-        if url.startswith('file://'):
-            url = url[len('file://'):]
+        if not directory:
+            raise Exception("'directory' must be a string")  # acutally Path would also work
+
+        directory = Path(directory)
+        directory.mkdir(parents=True, exist_ok=True)
+        directory = str(directory.resolve())
+
+        subdir_path = Path(subdir if subdir else '')
+        basename_path = Path(basename if basename else nth_url_segment(url))
+        ret = str(Path(subdir_path, basename_path))
+        dst_path = Path(directory, ret)
+
+        #  log.info("\n\tdst_path='%s \n\turl=%s", dst_path, url)
+        #  print('url=%s', url)
+        #  print('directory=%s', directory)
+        #  print('subdir_path=%s', subdir_path)
+        #  print('basename_path=%s', basename_path)
+        #  print('ret=%s', ret)
+        #  print('dst_path=%s', dst_path)
```

Comment:
Maybe remove print debugging?

Reply:
It's commented out but I foresee more need for print-debugging in the future. Unless it hurts your sense of code aesthetics too much, I'd prefer to keep them there for the time being.